zio-interop-cats/shared/src/main/scala/zio/stream/interop/FS2StreamSyntax.scala

```diff
@@ -40,37 +40,65 @@ final class FS2RIOStreamSyntax[R, A](private val stream: Stream[RIO[R, _], A]) {
   def toZStream[R1 <: R](queueSize: Int = 16): ZStream[R1, Throwable, A] =
     if (queueSize > 1) toZStreamChunk(queueSize) else toZStreamSingle
 
-  private def toZStreamSingle[R1 <: R]: ZStream[R1, Throwable, A] =
-    ZStream.managed {
-      for {
-        queue <- Queue.bounded[Take[Throwable, A]](1).toManaged[R](_.shutdown)
-        _     <-
-          (stream.evalTap(a => queue.offer(Take.single(a))) ++ fs2.Stream
-            .eval(queue.offer(Take.end)))
-            .handleErrorWith(e => fs2.Stream.eval(queue.offer(Take.fail(e))).drain)
-            .compile[RIO[R, _], RIO[R, _], Any]
-            .resource
-            .drain
-            .toManagedZIO
-            .fork
-      } yield ZStream.fromQueue(queue).flattenTake
-    }.flatten
-
-  private def toZStreamChunk[R1 <: R](queueSize: Int): ZStream[R1, Throwable, A] =
-    ZStream.managed {
-      for {
-        queue <- Queue.bounded[Take[Throwable, A]](queueSize).toManaged(_.shutdown)
-        _     <- {
-          stream
-            .chunkLimit(queueSize)
-            .evalTap(a => queue.offer(Take.chunk(zio.Chunk.fromIterable(a.toList))))
-            .unchunk ++ fs2.Stream.eval(queue.offer(Take.end))
-        }.handleErrorWith(e => fs2.Stream.eval(queue.offer(Take.fail(e))).drain)
-          .compile[RIO[R, _], RIO[R, _], Any]
-          .resource
+  def toZStreamChunk[R1 <: R](queueSize: Int = 16): ZStream[R1, Throwable, A] = {
+    def integrate(stream: fs2.Stream[RIO[R, _], A], zQueue: Queue[Take[Throwable, A]]): fs2.Pull[RIO[R, _], A, Unit] =
+      stream.pull.uncons.flatMap {
+        case None =>
+          fs2.Pull.eval(zQueue.offer(Take.end)) >> fs2.Pull.done
+
+        case Some((fs2Chunk: fs2.Chunk[A], stream)) =>
+          val offer = zQueue.offer(Take.chunk(toZioChunk(fs2Chunk)))
+          fs2.Pull.eval(offer) >> integrate(stream, zQueue)
+      }
+
+    convert(queueSize, integrate)
+  }
+
+  private def toZStreamSingle[R1 <: R]: ZStream[R1, Throwable, A] = {
+    def integrate(stream: fs2.Stream[RIO[R, _], A], zQueue: Queue[Take[Throwable, A]]): fs2.Pull[RIO[R, _], A, Unit] =
+      stream.pull.uncons.flatMap {
+        case None =>
+          fs2.Pull.eval(zQueue.offer(Take.end)) >> fs2.Pull.done
+
+        case Some((elements, stream)) =>
+          val offer = zQueue.offerAll(toZioChunk(elements).map(Take.single))
```

Comment:
I was not sure whether we could merge this with `Take.chunk(toZioChunk(fs2Chunk))` or is it better to use `toZioChunk(elements).map(Take.single)`?

Also, It might be better to use uncons1 in this single-element case because I feel like I'm building an bounded(chunk-size-of-fs2-stream) atop a single element queue

Reply:
I think you definitely want to offer a single `Take` containing a chunk of elements here.

Reply:
Hey @adamgfraser based on this, I can merge both implementations together since the previous chunk counterpart was doing that